crates/typst-syntax/src/ast.rs

```diff
@@ -1,6 +1,83 @@
+
+Raw nodes are parsed by recognizing paired backtick delimiters, which you will
+find as CST nodes with the [`RawDelim`] kind. However, the AST doesn't include
+these delimiters because it _abstracts_ over the backticks. Instead, the parent
+raw node will only use its child [`RawDelim`] CST nodes to determine whether the
+element is a block or inline.
+
+## The AST is Typed
+
+The AST gives access to methods on nodes of the tree. The [`Raw`] node provides
+3 methods that specify its abstract representation: [`Raw::lines()`] returns the
+raw text as an iterator of lines, [`Raw::lang()`] provides the optionally
+present [`RawLang`] language tag, and [`Raw::block()`] gives a bool for whether
+the raw element is a block or inline.
+
+These methods are implemented by assuming the structure of [`SyntaxNode`]s will
+match what is generated by the parser. The [`Raw`] struct really is just a
+wrapper over a [`SyntaxNode`] reference. But to construct the wrapper and access
+the methods, we must check that the [`SyntaxKind`] of the underlying node
+matches.
+
+And this is a safe interface because the Rust compiler only gives access to the
+methods after a [`Raw`] struct is created via [`AstNode::from_untyped`]. The
+[`node!`] macro implements `from_untyped` by checking the node's kind before
+constructing it, returning `Some()` only if the kind matches the desired struct.
+
+## The AST is Lazy
+
+Being lazy means that the untyped CST nodes are converted to typed AST Nodes
+only as the tree is traversed. If we parse a file and a raw block is contained
+in a branch of an if-statement that we don't take, then we won't pay the cost of
+creating an iterator over the lines or checking whether it was a block or
+inline (although it will still be parsed into nodes). This is also a factor of
+the current "tree-interpreter" evaluation model. A bytecode interpreter might
+instead eagerly convert the AST into bytecode, but it will still traverse using
+this lazy interface.
+
+## The AST is rooted in the [`Markup`] node
+
+Finally, the AST is rooted in the [`Markup`] node, which provides only one
+method: [`Markup::exprs`]. This returns an iterator of the main [`Expr`] enum.
+[`Expr`] is important because it contains the majority of expressions that Typst
+will evaluate. Not just markup, but also math and code expressions. Not all
+expression types are available from the parser at every step, but this does
+decrease the amount of wrapper enums needed in the AST, and this file is long
+enough already.
+
+Expressions also branch off into the remaining tree. You can view enums in this
+file as edges on a graph: areas where the tree has paths from one type to
+another (accessed through methods), then structs are the nodes of the graph,
+providing methods that return enums, etc. etc.
+
+---
+
+So that's the Typst AST interface!
+
+Feel free to ask questions in the contributors chat if anything isn't fully
+clear, or give us suggestions on ways to improve these docs!
```

Comment:
This seems like unnecessary fluff.
Also, referencing the contributors chat without saying where it is is going to confuse people that aren't already on the Discord server. Also also, as someone who doesn't really _like_ discord and its proprietary nature, I'd like to avoid linking to the discord from the source code.

Reply:
I agree that we should leave it out here, mostly because we don't want to add this all over the place. The CONTRIBUTING.md is the correct place for this and it has that info already.